File: go-backend-interview-prep/05_errors_and_testing/07_testing_basics/add_test.py
"""Tests for testing basics -- Python equivalent of add_test.go."""

import unittest
from main import add, abs_val, is_even


class TestAdd(unittest.TestCase):
    def test_positive(self):
        self.assertEqual(add(2, 3), 5)

    def test_zeros(self):
        self.assertEqual(add(0, 0), 0)

    def test_negative(self):
        self.assertEqual(add(-1, 1), 0)
        self.assertEqual(add(-3, -7), -10)


class TestAbs(unittest.TestCase):
    def test_negative(self):
        self.assertEqual(abs_val(-7), 7)

    def test_positive(self):
        self.assertEqual(abs_val(5), 5)

    def test_zero(self):
        self.assertEqual(abs_val(0), 0)


class TestIsEven(unittest.TestCase):
    def test_cases(self):
        cases = [
            (0, True),
            (1, False),
            (2, True),
            (-3, False),
            (-4, True),
        ]
        for n, want in cases:
            with self.subTest(n=n):
                self.assertEqual(is_even(n), want)


if __name__ == "__main__":
    unittest.main()

File: go-backend-interview-prep/05_errors_and_testing/07_testing_basics/add_test.go
package main

import "testing"

// assertEqual is a test helper that reports the caller's line on failure.
func assertEqual(t *testing.T, got, want int) {
	t.Helper()
	if got != want {
		t.Errorf("got %d, want %d", got, want)
	}
}

func TestAdd(t *testing.T) {
	assertEqual(t, Add(2, 3), 5)
	assertEqual(t, Add(0, 0), 0)
	assertEqual(t, Add(-1, 1), 0)
	assertEqual(t, Add(-3, -7), -10)
}

func TestAbs(t *testing.T) {
	assertEqual(t, Abs(-7), 7)
	assertEqual(t, Abs(5), 5)
	assertEqual(t, Abs(0), 0)
}

func TestIsEven(t *testing.T) {
	tests := []struct {
		input int
		want  bool
	}{
		{0, true},
		{1, false},
		{2, true},
		{-3, false},
		{-4, true},
	}

	for _, tc := range tests {
		got := IsEven(tc.input)
		if got != tc.want {
			t.Errorf("IsEven(%d) = %v, want %v", tc.input, got, tc.want)
		}
	}
}
